### PR TITLE
Api: Rework API for `Yatagan` loaders and reflection setup.

### LINKED ISSUE #30: Compose "api-*" artifacts contents (in terms of backend choice) more practical
As of now, there are the following "public api" artifacts:
1. `api-public` - just the annotations (API), no entry-point (EP - `Yatagan` object).
2. `api-compiled` - depends on `api-public`, adds `Yatagan` which loads compiled implementations.
3. `api-dynamic` - depends on `api-public`, adds `Yatagan` which loads dynamic implementations (can load compiled if configured) and comes with extra API for configuring reflection-specific things.

Now this "layout" seems to work, but apparently has some drawbacks. There's no real problem for the _application_ developers, as they can just use the correct backend-specific artifact and be done with it. But when it comes to _library_ development, some unpleasant things begin to surface: turns out it's not okay for library authors to depend on any backend-specific artifact, as they would actually force the backend usage on all the clients this way. One can't have multiple backends on the runtime classpath, as they would conflict. And the only backend-agnostic artifact is `api-public` which doesn't contain any `Yatagan` EPs, so there's no way for a library to create its own components in its code. If the library is not published and serves only as a way to organize code in the multi-module project, developers can get away with twiddling with tools like `compileOnly`/`runtimeOnly` and so on, but it is in no way a general solution.

So here I propose to rearrange the api/EP/loader code so its more practical to use and general integration logic can be formulated, maybe even with Gradle plugin. 

## New structure (draft)

1. `api:public` - contains both API and EP*. EP has the following behavior - they _try to locate a **dynamic loader** class_ (once, the result is cached - very fast). If the loader is present, EP delegates to it. If not - EP loads the compiled implementations.
2. `rt:loader` - the **dynamic loader** class and the reflection engine - loads and serves reflection-based implementations.

*It's not yet clear how to handle reflection-specific API - should they be included into `api:public` and be no-op by default and delegate to the `rt:loader`?

This way every project module can depend on `api:public` and its code will compile regardless. And then, depending on the usage and settings, additional configuration can be done - adding code generators or `rt:loader` as `runtimeOnly`. This way published libraries can depend on `api:public` and ship with generated implementations of their internal components. While clients can still include `rt:loader` and use reflection backend for themselves.

If the project depends on a library, that uses yatagan internally and the project itself chooses to use `rt:loader`, then we probably don't want for this to affect the library behavior. In other words, how exactly should `api:public`'s loader logic operate in regards to compiled implementations? Should it load them unconditionally if detected and try to use dynamic loader only if no compiled implementation is present? This works well in theory and preserves library behavior (if it comes with generated impls). But practice shows that build systems can leave stale generated code in the runtime classpath after a developer switched to reflection, e.g. via a build flag, then the loader would still use compiled impl instead of reflection. Need to give this some more thought.

### PR DESCRIPTION
**Pull request by Jeffset**
`Yatagan` EP is now defined a single time and sits in `api:public`, along with the implementation for loading compiled graphs. It can delegate to reflection backend (via `ImplementationLoader` interface)
 if it is present in the runtime classpath (`api:dynamic`).

Reflection-specific API is now a separate thing, `YataganReflection`,
 that sits in `:api:dynamic`.

`api:compiled` is now an empty artifact left for some degree of compatibility.

With these changes it's now possible to access the Yatagan EP without
 choosing the specific backend, i.m. loader always tries to load
 compiled implementations optionally falling back to reflection,
 if it's present in the runtime classpath.

As this design was implemented with library development in mind, it's
 now very important to keep `Yatagan` loader being able to load
 implementations generated by earlier versions of Yatagan's codegen.
 For that purpose a special `LoaderCompatibilityTest` is introduced.

Yatagan 1.2.0 already broke `create()` loader compatibility, so
 this CL fixes that with a workaround.

Closes #30

### CODE REVIEW

**Comment by Jeffset** [comment]
Will edit the docs in #20

File: api/public/src/main/kotlin/Yatagan.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        return autoBuilder(componentClass).create()
+    }
+
+    private inline fun <T> loadImpl(load: ImplementationLoader.() -> Result<T>): T {
+        return loaderForReflection?.let { reflection ->
+            // Always try to load generated and compiled implementation first.
+            // REASON: Libraries, that use Yatagan internally and ship with generated implementations
+            // (and doesn't do any shadowing/repackaging/bytecode rewriting) don't want their internal components
+            // to start using reflection if application developers use reflection.
+            loaderForCompiled.load().getOrElse {
```
Comment (Jeffset) [inline api/public/src/main/kotlin/Yatagan.kt:99]: Lost logging here, was [here](https://github.com/yandex/yatagan/pull/45/files#diff-26974fe3b232bf2722419ab91020545c38200e98d9cae1198ce54e5fe7cc37c1L156).

**Review by ArturVasilov — APPROVED** [review]

File: api/dynamic/src/main/kotlin/ReflectionLoader.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import com.yandex.yatagan.base.singleInitWithFallback
+import com.yandex.yatagan.internal.ImplementationLoader
+import com.yandex.yatagan.internal.YataganInternal
+import com.yandex.yatagan.rt.engine.RuntimeEngine
+import com.yandex.yatagan.rt.support.DynamicValidationDelegate
+import com.yandex.yatagan.rt.support.Logger
+
+/**
+ * Instantiated reflectively.
+ * WARNING: Keep class name in sync with [com.yandex.yatagan.Yatagan.DYNAMIC_LOADER_DELEGATE_CLASS_NAME].
```
Comment (byar) [inline api/dynamic/src/main/kotlin/ReflectionLoader.kt:29]: Test for that maybe?

**Review by byar — APPROVED** [review]

File: api/dynamic/src/main/kotlin/ReflectionLoader.kt (hunk 2 above)
Comment (Jeffset) [reply api/dynamic/src/main/kotlin/ReflectionLoader.kt:29]: It is already covered. Basically, every `[RT]` test with a `fun test()` will fail if this doesn't match.

**Comment by Jeffset** [comment]
Replaced 1.2.0-rc1 with a release version in tests.

**Pull request merged**